rasa/core/policies/flow_policy.py

```diff
@@ -372,202 +514,189 @@ def _is_step_completed(
-                return (FLOW_PREFIX + new_flow.id, [], 1.0)
-            else:
-                logger.debug("No startable flow found. Skipping prediction.")
-                return (None, [], 0.0)
+    ) -> Tuple[ActionPrediction, List[Event]]:
+        """Select the next action to execute.
 
-        # TODO: convert this to a proper step and transition model. right now
-        #  this is a bit of a mess with the flow state and the tracker state
+        Advances the current flow and returns the next action to execute. A flow
+        is advanced until it is completed or until it predicts an action. If
+        the flow is completed, the next flow is popped from the stack and
+        advanced. If there are no more flows, the action listen is predicted.
 
-        if not (current_step := self._get_current_step()) or not self.flow_state:
-            # If the next step is not set, we return the first step
-            # if there is one
-            next_step = self.start_flow(tracker, domain)
+        Args:
+            tracker: The tracker to get the next action for.
+            domain: The domain to get the next action for.
 
-        elif not self._is_step_completed(current_step, tracker):
-            # TODO: figure out
-            raise FlowException("Not quite sure what to do here yet.")
-        else:
-            # If the step is completed, we get the next step
-            next_step = self._get_next_step(
-                tracker, domain, current_step, self.flow_state.flow_id
-            )
+        Returns:
+            The next action to execute, the events that should be applied to the
+            tracker and the confidence of the prediction."""
+
+        predicted_action: Optional[ActionPrediction] = None
+        gathered_events = []
+
+        while not predicted_action:
+            if not (current_flow := self.flow_stack.top_flow(self.all_flows)):
+                # If there is no current flow, we assume that all flows are done
+                # and there is nothing to do. The assumption here is that every
+                # flow ends with an action listen.
+                predicted_action = ActionPrediction(ACTION_LISTEN_NAME, 1.0)
+                break
+
+            if not (previous_step := self.flow_stack.top_flow_step(self.all_flows)):
+                raise FlowException(
+                    "The current flow is set, but there is no current step. "
+                    "This should not happen, if a flow is started it should be set "
+                    "to __start__ if it ended it should be popped from the stack."
+                )
 
-        if next_step:
-            action, events = self._get_action_for_next_step(next_step, tracker, domain)
-            return (action, events, 1.0)
+            if not self._is_step_completed(previous_step, tracker):
+                # TODO: figure out
+                raise FlowException("Not quite sure what to do here yet.")
```

Comment:
might be an opportunity to get rid of the loop action? If a question isn't answered yet, ask it again?

Reply:
yes i think we should consider it - wouldn't want to go there just now though